cartography/intel/aws/ec2/tgw.py

```diff
@@ -0,0 +1,214 @@
+    tgw_attachments = []
+    for page in paginator.paginate():
+        tgw_attachments.extend(page['TransitGatewayAttachments'])
+    return {"TransitGatewayAttachments": tgw_attachments}
+
+
+@timeit
+def get_tgw_vpc_attachments(boto3_session, region):
+    client = boto3_session.client('ec2', region_name=region, config=get_botocore_config())
+    paginator = client.get_paginator('describe_transit_gateway_vpc_attachments')
+    tgw_vpc_attachments = []
+    for page in paginator.paginate():
+        tgw_vpc_attachments.extend(page['TransitGatewayVpcAttachments'])
+    return {"TransitGatewayVpcAttachments": tgw_vpc_attachments}
+
+
+@timeit
+def load_transit_gateways(neo4j_session, data, region, current_aws_account_id, aws_update_tag):
+    ingest_transit_gateway = """
+    MERGE (ownerAccount:AWSAccount {id: {OwnerId}})
+    ON CREATE SET ownerAccount.firstseen = timestamp(), ownerAccount.foreign = true
+    SET ownerAccount.lastupdated = {aws_update_tag}
+
+    MERGE (tgw:AWSTransitGateway {id: {TgwId}, arn: {ARN}})
+    ON CREATE SET tgw.firstseen = timestamp()
+    SET tgw.ownerid = {OwnerId}, tgw.state = {State},
+    tgw.description = {Description},
+    tgw.region = {Region},
+    tgw.lastupdated = {aws_update_tag}
+
+    MERGE (ownerAccount)-[r:RESOURCE]->(tgw)
+    ON CREATE SET r.firstseen = timestamp()
+    SET r.lastupdated = {aws_update_tag}
+    """
+
+    for tgw in data["TransitGateways"]:
+        tgw_id = tgw["TransitGatewayId"]
+
+        neo4j_session.run(
+            ingest_transit_gateway,
+            TgwId=tgw_id,
+            ARN=tgw["TransitGatewayArn"],
+            Description=tgw.get("Description"),
+            Region=region,
+            AWS_ACCOUNT_ID=current_aws_account_id,
+            OwnerId=tgw["OwnerId"],
+            State=tgw["State"],
+            aws_update_tag=aws_update_tag,
+        )
+        _attach_shared_transit_gateway(
+            neo4j_session, tgw, region, current_aws_account_id, aws_update_tag,
+        )
+
+
+@timeit
+def _attach_shared_transit_gateway(neo4j_session, tgw, region, current_aws_account_id, aws_update_tag):
+    attach_tgw = """
+    MERGE (tgw:AWSTransitGateway {id: {TransitGatewayId}})
+    ON CREATE SET tgw.firstseen = timestamp()
+    SET tgw.lastupdated = {aws_update_tag}
+
+    WITH tgw
+    MATCH (currentAccount:AWSAccount{id: {AWS_ACCOUNT_ID}})
+    MERGE (tgw)-[s:SHARED_WITH]->(currentAccount)
+    ON CREATE SET s.firstseen = timestamp()
+    SET s.lastupdated = {aws_update_tag}
+    """
+
+    if tgw["OwnerId"] != current_aws_account_id:
+        neo4j_session.run(
+            attach_tgw,
+            TransitGatewayId=tgw["TransitGatewayId"],
+            AWS_ACCOUNT_ID=current_aws_account_id,
+            aws_update_tag=aws_update_tag,
+        )
+
+
+@timeit
+def load_tgw_attachments(neo4j_session, data, region, current_aws_account_id, aws_update_tag):
+    ingest_transit_gateway = """
+    MERGE (tgwa:AWSTransitGatewayAttachment{id: {TgwAttachmentId}})
```

Comment:
q: are we sure TgwAttachementId will be unique across AWS accounts? unless specified in the api doc, I would assume not.

Reply:
I'm going to assume that it is, and even if it isn't, it would be astronomically unlikely that we'd have an overlap in our domain of inter-connected AWS accounts. We've built all of our infrastructure so far with this assumption and haven't had any problems, so I hope I'm right!

Reply:
could be, image some that are merging 20-30K accounts like my team aims for now. Collisions of that kind are very hard to spot and often found during incident/urgency.

Reply:
20-30K 😮 how long does cartography take to run on such a huge deployment?

I really think it's exceedingly unlikely to find a collision among roughly 10e+4 objects with a 17-character hex ID providing ~ 10e+20 unique IDs.
The TGW attachment API does not return an ARN but I suppose if we really wanted to we can construct one from all given info, e.g., `f'arn:aws:ec2:{aws_region}:{ResourceOwnerId}:transit-gateway-attachment/tgw-attach-0f0f0f0f0f0f0f0f0`
(side note, I didn't find any reference to the ARN structure for a TGW attachment in AWS API docs, so I made one up)

Reply:
Tks for clarifying so we can rely on it for the id field. For scanning large # of account, it requires a different pipeline, you can sync them in parallel but Cartography doesn't support it. Might be something we contribute back.